File: 538.把二叉搜索树转换为累加树.js
/*
 * @lc app=leetcode.cn id=538 lang=javascript
 *
 * [538] 把二叉搜索树转换为累加树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
let pre;
const convBST = (node) => {
  if (!node) return;
  convBST(node.right);
  if (pre) node.val += pre.val;
  pre = node;
  convBST(node.left);
};
var convertBST = function (root) {
  // 递归
  // pre = null;
  // convBST(root);
  // return root;
  // 迭代
  let pre,
    stack = [],
    top = 0;
  stack[top++] = root;
  while (top) {
    let node = stack[--top];
    if (node === null) {
      node = stack[--top];
      if (pre) node.val += pre.val;
      pre = node;
    } else {
      if (node && node.left) stack[top++] = node.left;
      stack[top++] = node;
      stack[top++] = null;
      if (node && node.right) stack[top++] = node.right;
    }
  }
  return root;
};
// @lc code=end

File: store/538.把二叉搜索树转换为累加树.c
/*
 * @lc app=leetcode.cn id=538 lang=c
 *
 * [538] 把二叉搜索树转换为累加树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
// int total;
// struct TreeNode *convert(struct TreeNode *root)
// {
//   if (root == NULL)
//     return root;
//   convert(root->right);
//   root->val += total;
//   total = root->val;
//   convert(root->left);
//   return root;
// }

struct TreeNode *convertBST(struct TreeNode *root)
{
  // 法一：递归
  // total = 0;
  // return convert(root);
  // 法二：迭代
  struct TreeNode **stack, *node;
  int top, total;
  top = total = 0;
  stack = (struct TreeNode **)malloc(sizeof(struct TreeNode *) * 1000);
  if (root != NULL)
  {
    stack[top++] = root;
    while (top > 0)
    {
      node = stack[--top];
      if (node != NULL)
      {
        if (node->left != NULL)
          stack[top++] = node->left;
        stack[top++] = node;
        stack[top++] = NULL;
        if (node->right != NULL)
          stack[top++] = node->right;
      }
      else
      {
        node = stack[--top];
        node->val += total;
        total = node->val;
      }
    }
  }
  return root;
}
// @lc code=end
